src/compute/src/render/iteration.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::any::Any;
+use std::collections::{BTreeMap, BTreeSet};
+use std::rc::Rc;
+use std::sync::Arc;
+
+use differential_dataflow::AsCollection;
+use timely::communication::Allocate;
+use timely::dataflow::scopes::Child;
+use timely::dataflow::Scope;
+use timely::order::Product;
+use timely::worker::Worker as TimelyWorker;
+
+use mz_compute_client::plan::Plan;
+use mz_compute_client::types::dataflows::{DataflowDescription, IndexDesc};
+use mz_expr::Id;
+use mz_ore::collections::CollectionExt as IteratorExt;
+use mz_repr::{GlobalId, Row};
+use mz_storage_client::controller::CollectionMetadata;
+use mz_storage_client::source::persist_source;
+
+pub use super::context::CollectionBundle;
+use super::context::{ArrangementFlavor, Context};
+use crate::arrangement::manager::TraceBundle;
+use crate::compute_state::ComputeState;
+use crate::render::RenderTimestamp;
+
+pub fn build_compute_dataflow<A: Allocate>(
+    timely_worker: &mut TimelyWorker<A>,
+    compute_state: &mut ComputeState,
+    dataflow: DataflowDescription<Plan, CollectionMetadata>,
+) {
+    let worker_logging = timely_worker.log_register().get("timely");
+    let name = format!("Dataflow: {}", &dataflow.debug_name);
+
+    timely_worker.dataflow_core(&name, worker_logging, Box::new(()), |_, scope| {
+        // The scope.clone() occurs to allow import in the region.
+        // We build a region here to establish a pattern of a scope inside the dataflow,
+        // so that other similar uses (e.g. with iterative scopes) do not require weird
+        // alternate type signatures.
+        scope.clone().iterative::<usize, _, _>(|region| {
+            let mut context = crate::render::context::Context::for_dataflow(
+                &dataflow,
+                scope.addr().into_element(),
+            );
+            let mut tokens = BTreeMap::new();
+
+            // Import declared sources into the rendering context.
+            for (source_id, (source, _monotonic)) in dataflow.source_imports.iter() {
+                let mut mfp = source.arguments.operators.clone().map(|ops| {
+                    mz_expr::MfpPlan::create_from(ops)
+                        .expect("Linear operators should always be valid")
+                });
+
+                // Note: For correctness, we require that sources only emit times advanced by
+                // `dataflow.as_of`. `persist_source` is documented to provide this guarantee.
+                let (mut ok_stream, err_stream, token) = persist_source::persist_source(
+                    scope,
```

Comment:
Here we are passing `scope` to `persist_source` instead of `region` as in the non-iterative case. Is this to avoid exposing two-dimensional timestamps and the product type to the `persist` part of the code? If so, would it be possible to make the two code paths uniform by passing `scope` in the non-iterative implementation as well?

Reply:
I did this! At least, built regions for inputs and everything just up until index importing. There is still the opportunity to tidy it up (it leaves a big `if .. else ..` block there, but getting the seven lifetime'd arguments sorted out seemed most likely to mess up understanding at this moment (while we are assessing its viability).

Reply:
Super, thanks! I captured a potential follow-up after this work stabilizes in https://github.com/MaterializeInc/materialize/issues/16685.